test/lit/passes/reorder-globals.wast

```diff
@@ -0,0 +1,376 @@
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (global.get $a)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  ;; RTRIP:      (func $uses
+  ;; RTRIP-NEXT:  (global.set $b
+  ;; RTRIP-NEXT:   (i32.const 30)
+  ;; RTRIP-NEXT:  )
+  ;; RTRIP-NEXT:  (global.set $b
+  ;; RTRIP-NEXT:   (i32.const 40)
+  ;; RTRIP-NEXT:  )
+  ;; RTRIP-NEXT:  (drop
+  ;; RTRIP-NEXT:   (global.get $a)
+  ;; RTRIP-NEXT:  )
+  ;; RTRIP-NEXT: )
+  (func $uses
+    (global.set $b
+      (i32.const 30)
+    )
+    (global.set $b
+      (i32.const 40)
+    )
+    (drop
+      (global.get $a)
+    )
+  )
+)
+
+;; As above, but flipped so now $a has more, and should remain first.
+(module
+  ;; CHECK:      (global $a (mut i32) (i32.const 10))
+  ;; RTRIP:      (global $a (mut i32) (i32.const 10))
+  (global $a (mut i32) (i32.const 10))
+  ;; CHECK:      (global $b (mut i32) (i32.const 20))
+  ;; RTRIP:      (global $b (mut i32) (i32.const 20))
+  (global $b (mut i32) (i32.const 20))
+
+  ;; CHECK:      (func $uses
+  ;; CHECK-NEXT:  (global.set $a
+  ;; CHECK-NEXT:   (i32.const 30)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (global.set $a
+  ;; CHECK-NEXT:   (i32.const 40)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (global.get $b)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  ;; RTRIP:      (func $uses
+  ;; RTRIP-NEXT:  (global.set $a
+  ;; RTRIP-NEXT:   (i32.const 30)
+  ;; RTRIP-NEXT:  )
+  ;; RTRIP-NEXT:  (global.set $a
+  ;; RTRIP-NEXT:   (i32.const 40)
+  ;; RTRIP-NEXT:  )
+  ;; RTRIP-NEXT:  (drop
+  ;; RTRIP-NEXT:   (global.get $b)
+  ;; RTRIP-NEXT:  )
+  ;; RTRIP-NEXT: )
+  (func $uses
+    (global.set $a
+      (i32.const 30)
+    )
+    (global.set $a
+      (i32.const 40)
+    )
+    (drop
+      (global.get $b)
+    )
+  )
+)
+
+;; $b has more uses, but it depends on $a and cannot be sorted before it.
+(module
+  ;; CHECK:      (global $a i32 (i32.const 10))
+  ;; RTRIP:      (global $a i32 (i32.const 10))
+  (global $a i32 (i32.const 10))
+  ;; CHECK:      (global $b i32 (global.get $a))
+  ;; RTRIP:      (global $b i32 (global.get $a))
+  (global $b i32 (global.get $a))
+
+  ;; CHECK:      (func $uses
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (global.get $b)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  ;; RTRIP:      (func $uses
+  ;; RTRIP-NEXT:  (drop
+  ;; RTRIP-NEXT:   (global.get $b)
+  ;; RTRIP-NEXT:  )
+  ;; RTRIP-NEXT: )
+  (func $uses
+    (drop
+      (global.get $b)
+    )
+  )
+)
+
+;; $c has more uses, but it depends on $b and $a and cannot be sorted before
+;; them. Likewise $b cannot be before $a.
+(module
+  ;; CHECK:      (global $a i32 (i32.const 10))
+  ;; RTRIP:      (global $a i32 (i32.const 10))
+  (global $a i32 (i32.const 10))
+  ;; CHECK:      (global $b i32 (global.get $a))
+  ;; RTRIP:      (global $b i32 (global.get $a))
+  (global $b i32 (global.get $a))
+  ;; CHECK:      (global $c i32 (global.get $b))
+  ;; RTRIP:      (global $c i32 (global.get $b))
+  (global $c i32 (global.get $b))
```

Comment:
This is not actually valid Wasm because constant expressions can only refer to imported globals. Is that a problem?

Reply:
Right, I forgot that spec limitation... that does make the problem simpler I guess, and the pass is more general than it needs to be. But there's no harm in supporting more things, and I'm sure the spec will eventually allow this, at which point this code will just work.